### PR TITLE
Add graph debug function of LR and step

### PR DESCRIPTION
**Pull request by lixiang007666**
This PR is done:

- [x] Graph adds the debug function of outputting LR and step(debug(v_level>=1)).
- [x] Only output this information at rank 0.
- [x] Add debug LR docstr.

**This function is mainly for debug.**

Related issue：https://github.com/Oneflow-Inc/OneTeam/issues/1123

~~usage:~~
~~export ONEFLOW_DEBUG_GRAPH_STEP_LR=1~~
The above way of using environment variables has been cancelled.

output(When debug(v_level>=1)):

image
image

image

### CODE REVIEW

File: python/oneflow/nn/graph/graph.py [hunk 1]
```diff
@@ -513,6 +513,13 @@ def __print(self, s_level=2, v_level=0, msg: str = ""):
             if (s_level > 0) or (s_level == 0 and v_level <= self._debug_max_v_level):
                 print(msg, flush=True)
 
+    # def _print_lr(self):
```
Comment (strint) [inline python/oneflow/nn/graph/graph.py:516]: 这个是无用的，删掉吧

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 2]
```diff
@@ -287,6 +287,8 @@ void LearningRateScheduleKernel::ForwardDataContent(KernelContext* ctx) const {
   if (conf.has_learning_rate_decay()) {
     learning_rate = GetDecayedLearningRate(conf.learning_rate_decay(), learning_rate, train_step);
   }
+  std::cout << "Last step " << train_step << " adjusting learning rate to " << learning_rate
+            << std::endl;
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:291]: 这个打印的效果可以看下。

看起来这里直接有train step的信息，太好了。不用折腾。

Comment (lixiang007666) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp]: image

现在设置的是每100个step输出一次，是否可以用环境变量或开发一个api，控制是否输出lr、step信息和设置step输出的间隔。

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 3]
```diff
@@ -287,6 +287,10 @@ void LearningRateScheduleKernel::ForwardDataContent(KernelContext* ctx) const {
   if (conf.has_learning_rate_decay()) {
     learning_rate = GetDecayedLearningRate(conf.learning_rate_decay(), learning_rate, train_step);
   }
+  if (std::getenv("ONEFLOW_DEBUG_GRAPH_STEP_LR") != nullptr) {
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:290]: 想了一下感觉用环境变量还是一个太简陋的做法，还是做到nn.Graph上比较好，最好复用graph的lr对象的verbose参数。

graph的编译是同步操作，所以我们可以比较容易的用scope去做一些开关。但是graph的执行是异步的，这时scope就无法控制kernel的行为了。

这时指引我们必须改KernelContext了，把debug开关通过KernelContext传递进来。这样：
- python对象控制，可以做到单graph的单rank级别
- 接口体验和torch一致

这两点是环境变量无法达到到的。

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@ class LearningRateScheduleKernel final : public Kernel {
       log_stream_ = TeePersistentLogStream::Create("train_step2lr.csv");
       (*log_stream_) << "train_step, lr\n";
     }
+    // if (FLAGS_v >= 1) { print_step_lr_ = true; }
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: 用nn.Graph.debug(1)会控制FLAGS_v，FLAGS_v给print_step_lr_一个状态，作为kernel的状态，就可以使得kernel后面总是用这个状态工作了

Comment (chengtbf) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: 怎么定义 “编译时” 呢？ 😂   以前 Lazy 中的编译时表示的是 从 Job -> Plan 的过程。  Runtime 表示的是从 Plan 到 Actor Graph 的过程。  因此 Actor、 Kernel 都是运行时概念；  Task、Op 是编译时概念。

nn.Graph.debug() 可以控制是指的？ 如果 debug 可以控制到 第一次真正运行之前，那么是可以控制 Kernel init 的。

Comment (strint) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: > 怎么定义 “编译时” 呢
这里说的的确有点模糊，具体就只之在nn.Graph._compile()函数里面的、同步调用的，都可以用一个作用域来控制，读取到nn.Graph.debug()的debug等级。

Comment (strint) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: nn.Graph其实还无法知道Thread里面的kernel真正init了是吧

Comment (strint) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: 我猜这里是work的，但是不一定是正确的，取决于actor执行Thread::PollMsgChannel的速度，但是刚初始化时，应该是很快的。

Comment (chengtbf) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: kernel init 如果是 construct actor 就会做（而不是 actor 第一次执行时）的话，那么这个就是 work 的，因为 https://github.com/Oneflow-Inc/oneflow/blob/f8479270792b3fedd63667bb26bb08f977a24b40/oneflow/core/job/runtime.cpp#L94 这里会同步等待 actor 都构造完才会返回 runtime 构造完毕。

Comment (strint) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:35]: 看了下，有这个`WaitUntilCntEqualZero("constructing_actor_cnt")`的确保证了是work的，actor和kernel都初始化好了。

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 5]
```diff
@@ -32,10 +33,13 @@ class LearningRateScheduleKernel final : public Kernel {
       log_stream_ = TeePersistentLogStream::Create("train_step2lr.csv");
       (*log_stream_) << "train_step, lr\n";
     }
-    // if (FLAGS_v >= 1) { print_step_lr_ = true; }
+
+    const int64_t FLAGS_v = oneflow::GetFLAGS_v().GetOrThrow();
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:37]: 这里直接类似GetFLAGS_v函数，直接读FLAGS_v就好

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 6]
```diff
@@ -288,8 +292,8 @@ void LearningRateScheduleKernel::ForwardDataContent(KernelContext* ctx) const {
   if (conf.has_learning_rate_decay()) {
     learning_rate = GetDecayedLearningRate(conf.learning_rate_decay(), learning_rate, train_step);
   }
-  // nn.Graph.debug(1) will print step and lr
-  if (print_step_lr_) {
+  // NOTE(lixiang): nn.Graph.debug(1) will print step and lr.
+  if (print_step_lr_ && train_step != 0) {
     std::cout << "Last step " << train_step << " adjusting learning rate to " << learning_rate
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:297]: 为什么不打印train_step0呢

**Comment by strint** [comment]
另外增加一个测试用例吧

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp (hunk 6 above)
Comment (lixiang007666) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:297]: 因为我发现，在`add_optimizer(optimizer,lr_sch=scheduler)`时，会打印python层Graph的LR对象的step0信息，就像这个PR最上面的描述。

Comment (strint) [reply oneflow/core/kernel/learning_rate_schedule_kernel.cpp:297]: 哦，是因为那个LR开了verbose，你关闭掉试试

File: oneflow/core/kernel/learning_rate_schedule_kernel.cpp [hunk 7]
```diff
@@ -34,7 +34,7 @@ class LearningRateScheduleKernel final : public Kernel {
       (*log_stream_) << "train_step, lr\n";
     }
 
-    const int64_t FLAGS_v = oneflow::GetFLAGS_v().GetOrThrow();
+    const int32_t FLAGS_v = oneflow::GetFLAGS_v().GetOrThrow();
```
Comment (strint) [inline oneflow/core/kernel/learning_rate_schedule_kernel.cpp:37]: FLAGS_v 不用专门获取，是glog，按说是直接可以用

**Review by strint — APPROVED** [review]

**Review by chengtbf — APPROVED** [review]
